lib/puppet/provider/service/systemd.rb

```diff
@@ -28,12 +28,33 @@ def disable
 
   def enabled?
     begin
-      systemctl("is-enabled", @resource[:name])
+      systemctl_info = systemctl(
+         'show',
+         @resource[:name],
+         '--property', 'LoadState',
+         '--property', 'UnitFileState',
+         '--no-pager'
+      )
+
+      svc_info = Hash.new
+      systemctl_info.split.map{|svc|
+        entry_pair = svc.split('=')
+        svc_info[entry_pair.first.to_sym] = entry_pair.last
+      }
+
+      # The masked state is equivalent to the disabled state in terms of
+      # comparison so we only care to check if it is masked if we want to keep
+      # it masked.
+      #
+      # We only return :mask if we're trying to mask the service. This prevents
+      # flapping when simply trying to disable a masked service.
+      return :mask if (@resource[:enable] == :mask) && (svc_info[:LoadState] == 'masked')
+      return :true if svc_info[:UnitFileState] == 'enabled'
     rescue Puppet::ExecutionFailure
-      return :false
+      # Don't worry about this failing, just return :false if it does.
```

Comment:
There are several failure cases that could happen when calling the systemctl command. This ignores those cases since they are not actually relevant to the state of the resource.